desc/compute/utils.py

```diff
@@ -620,20 +612,93 @@ def tupleset(t, i, value):
+        You may want to multiply q by the line length Jacobian.
+
+        Correctness is not guaranteed on grids with duplicate nodes.
+        An attempt to print a warning is made if the given grid has duplicate
+        nodes and is one of the predefined grid types
+        (Linear, Concentric, Quadrature).
+        If the grid is custom, no attempt is made to warn.
+
+    Parameters
+    ----------
+    grid : Grid
+        Collocation grid containing the nodes to evaluate at.
+    q : ndarray
+        Quantity to integrate.
+    line_label : str
+        The coordinate curve of rho, theta, or zeta to compute the integration over.
+        To clarify, a theta (poloidal) curve is the intersection of a
+        rho (flux) surface and zeta (toroidal) surface.
+    fix_surface : str, float
+        A tuple of the form: label, value.
+        fix_surface label should differ from line_label.
+        By default, fix_surface is chosen to be the flux surface at rho=1.
+
+    Returns
+    -------
+    integrals : ndarray
+        Line integrals of q over curves covering the given surface.
+
+    """
+    assert (
+        line_label != fix_surface[0]
+    ), "There is no valid use for this combination of inputs."
+    assert line_label == "theta" or not isinstance(
+        grid, ConcentricGrid
+    ), "ConcentricGrid should only be used for theta line integrals."
+    if isinstance(grid, LinearGrid) and grid.endpoint:
+        warnings.warn(
+            colored(
+                "Correctness not guaranteed on grids with duplicate nodes.", "yellow"
+            )
+        )
+    # Generate a new quantity q_prime which is zero everywhere
+    # except on the fixed surface, on which q_prime takes the value of q.
+    # Then forward the computation to surface_integrals.
+    # The differential element of the line integral, denoted dl,
+    # should correspond to the line label's spacing.
+    # The differential element of the surface integral is
+    # ds = dl * fix_surface_dl, so we scale q_prime by 1 / fix_surface_dl.
+    labels = {"rho": 0, "theta": 1, "zeta": 2}
+    column_id = labels[fix_surface[0]]
+    mask = grid.nodes[:, column_id] == fix_surface[1]
+    q_prime = (mask * jnp.atleast_1d(q).T / grid.spacing[:, column_id]).T
+    (surface_label,) = labels.keys() - {line_label, fix_surface[0]}
+    return surface_integrals(grid, q_prime, surface_label)
+
+
+def surface_integrals(grid, q=jnp.array([1.0]), surface_label="rho"):
     """Compute the surface integral of a quantity for all surfaces in the grid.
 
+    Notes
+    -----
+        It is assumed that the integration surface has area 4π^2 when the
+        surface label is rho and area 2π when the surface label is theta or
+        zeta. You may want to multiply q by the surface area Jacobian.
+
     Parameters
     ----------
     grid : Grid
         Collocation grid containing the nodes to evaluate at.
     q : ndarray
```

Comment:
Should update this with some explanation of the expected shapes of the inputs for function or vector valued integrands

Reply:
Updated. In any case the only requirement is that the first dimension has size grid.num_nodes as usual.